CI improvements

FlyGoat opened the pull request:
Convert to use toolchain built with GitHub action. (Despite sh and msp430 with reserved issue as mentioned in #900).

Wire up or1k and lm32 build tests.

@keith-packard Do you mind to fork https://github.com/FlyGoat/picolibc-ci-tools into picolibc organisation and tag a release so we can maintain it sustainably.

FlyGoat commented:
I'll try to look into msp430 issue and possibly or1k semihosting :-)

keith-packard commented:
Is there anything left here which needs to get pulled in?

FlyGoat commented:
Rebased, however it seems like CI is failing due to gcc-riscv64-unknown-elf missing from debian repo.

keith-packard commented on .github/Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM debian:testing
+FROM debian:sid
```
yeah, I know testing is busted today. I switched to it because unstable would be unusable sometimes. testing should be fixed in a couple of days once gcc-14 migrates.

FlyGoat replied:
> yeah, I know testing is busted today. I switched to it because unstable would be unusable sometimes. testing should be fixed in a couple of days once gcc-14 migrates.

My personal experience is sid tends to be more stable, I'll drop this commit.

keith-packard replied:
yeah, testing is having a rough time right now with gcc-14 taking a couple of months to migrate due to testing regressions. I guess if it doesn't recover soon, we're gonna need to switch to unstable or not have CI at all.

FlyGoat commented:
@keith-packard Rebased and passing all CI test :-)

keith-packard commented:
I went ahead and added semihosting support for lm32 for the binutils simulator, although that required hacks to the simulator to pass through all of the semihosting calls rather than only supporting exit. With that, I was able to run the tests. By building in debug mode, I was able to get everything to work except for printf-tests, and that uncovered a gcc bug in handling va_arg. Once I fixed that (a rather long adventure), I was able to get the whole test suite working, but only with `--buildtype=debug`. With any optimization at all, things fail pretty catastrophically.

binutils: https://github.com/picolibc/binutils-gdb/tree/lm32-syscalls
gcc: https://github.com/keith-packard/gcc/tree/lm32-va-arg
picolibc: https://github.com/picolibc/picolibc/tree/lm32

FlyGoat commented:
> I went ahead and added semihosting support for lm32 for the binutils simulator, although that required hacks to the simulator to pass through all of the semihosting calls rather than only supporting exit. With that, I was able to run the tests. By building in debug mode, I was able to get everything to work except for printf-tests, and that uncovered a gcc bug in handling va_arg. Once I fixed that (a rather long adventure), I was able to get the whole test suite working, but only with `--buildtype=debug`. With any optimization at all, things fail pretty catastrophically.
> 
> binutils: https://github.com/picolibc/binutils-gdb/tree/lm32-syscalls gcc: https://github.com/keith-packard/gcc/tree/lm32-va-arg picolibc: https://github.com/picolibc/picolibc/tree/lm32

Thanks! integrated patches to this PR: https://github.com/picolibc/picolibc-ci-tools/pull/1

FlyGoat commented:
Suppressed by #914,
